### ISSUE TITLE
Add example of custom serialization

### ISSUE THREAD

**Issue by mrzv** [avl/savefile#15]
It would be great to have an example of a custom serialization, using `WithSchema`, `Serialize`, and `Deserialize`. I'd like to serialize an enum that has a `PathBuf` in it, but it's tricky to figure out how this is supposed to work.

**Comment by avl** [comment]
Hello!

Because of the so-called 'orphan rules', it isn't really possible for some other crate to implement WithSchema etc for PathBuf. It can only be implemented in std or savefile itself.

What you can do is implement the Savefile-traits for a wrapper around PathBuf, and use that instead.

PathBuf is a bit tricky, since it can contain non-UTF8-characters, and might be OS-dependent. I've gone ahead and added an implementation for PathBuf, which basically assumes paths are valid UTF8. Hopefully this is good enough for most practical uses. Savefile tries to be OS agnostic, so it seems sort of reasonable to only support valid UTF-8 and not whatever raw byte arrays are supported on certain operating systems.

Anyway, in case you're curious, an implementation of the Savefile-traits for PathBuf would look something like this:

````rust
impl WithSchema for PathBuf {
    fn schema(_version: u32) -> Schema {
        Schema::Primitive(SchemaPrimitive::schema_string)
    }
}
impl Serialize for PathBuf {
    fn serialize<'a>(&self, serializer: &mut Serializer<'a>) -> Result<(), SavefileError> {
        let as_string : String = self.to_string_lossy().to_string();
        as_string.serialize(serializer)
    }
}
impl Deserialize for PathBuf {
    fn deserialize(deserializer: &mut Deserializer) -> Result<Self, SavefileError> {
        Ok(PathBuf::from(String::deserialize(deserializer)?))
    }
}
````

I've added this to the Savefile lib. 

Please tell me if 'to_string_lossy' is a bad idea.

I've also added a small example in the manual on how to implement Savefile-traits for a small custom struct.

The best way to learn how to implement the Savefile-traits is probably to read the Savefile source code - it already has implementations for many types.

One last thing - PR:s adding implementations for more types are welcome!

**Comment by mrzv** [comment]
Thanks for the explanation and for adding `PathBuf`. `to_string_lossy` is perfect for my use-case. My plan was to write a custom serializer for my enum and to convert the `PathBuf` to a string on the fly. The `serialize` and `deserialize` make perfect sense; it's the `Schema` I was struggling with, and still don't understand how it's supposed to work in general.

I hear you on the source code and on the PRs.

Thanks again.

**Comment by avl** [comment]
Right! I hear you.

The docs don't really go into depth on how the Schema-thing is supposed to work, especially for custom serializers. There is room for improvement here!

The basic point is that there are a few predefined schema types, mapping to the common types that savefile-derive supports and built-in types in the savefile library. The types are composable, forming a tree that is supposed to map to your serialized data. Savefile doesn't support cycles, neither in the data to serialize nor in the schema. For example, the Struct schema type has many fields, each of which could itself be a struct (or any other type) and so on.

The docs can be found here: https://docs.rs/savefile/0.7.2/savefile/enum.Schema.html

Going through them from the top, we have:

Struct - Something with named fields which may be of different types. Good for any type of aggregate of other objects. For example, let's say in your example you wanted to serialize a PathBuf and a mime-type . Then you might want to let the schema be a struct with two fields, one carrying the PathBuf and one carrying the mime-type. Note that each field in a Struct can have an arbitrary type itself, including Struct.

Enum - Something with many named variants. Each variant may have fields, and the fields may have different types.

Primitive - The standard integers, and strings

Vector - A sequence of elements which are all of the same type

Array - Like vector, but with a statically known size.

SchemaOption - A special case for Option. 

Undefined - Schema isn't specified

ZeroSize - Anything which has a zero-size in memory and is thus not actually serialized.

The basic building blocks that are right in most cases are Struct, Enum, Primitive and Vector.

Note that by default the schema itself is serialized to the file. When loading, the schema of the datatype being loaded into is calculated, and compared with the stored schema. If they do not match, an error is emitted.

The basic point of the schema-system is that if you accidentally change a variable from i16 to i32 (without correctly using the version management functions) you will still not get data-corruption, but a hard error with a (hopefully) understandable error message.

**Comment by avl** [comment]
Closing this now. PathBuf support will come in the next release (and is available on master to anyone who needs it before that!).

**Comment by mrzv** [comment]
I just tried using this in my actual code and got an error about `Introspect` not being implemented for `PathBuf`.

**Comment by avl** [comment]
Oh, I'm very sorry! I didn't really test the PathBuf implementation, which I should have done.

The default Savefile-derive also implements Introspect. And to do this, all parts of the datastructure must also implement Introspect.

But I forgot to add an implementation for Introspect to PathBuf, which means it only works for serialization/deserialization, not introspection. But that means the regular 'Savefile' derive can't be used on things containing PathBuf-objects.

**Comment by avl** [comment]
I'll fix this immediately!

**Comment by avl** [comment]
Please try again now, it should work.

**Comment by avl** [comment]
I've also published a new release. So just say savefile = "0.7.4" and I think (hope!) everything should work.

**Comment by mrzv** [comment]
Yup, it works. Well, more accurately `cargo check` passes. Thanks again!

**Comment by mrzv** [comment]
And the actual code does what it should. So the issues is solved.

I have one more unrelated problem. I can open a new issue, if you like, or maybe it's a "wontfix" situation. It seems `SavefileError` doesn't implement `std::error::Error` trait. I'm using [eyre](https://docs.rs/eyre/0.6.0/eyre/) in my code, but it refuses to work with function that return `SavefileError`. (I'm guessing the trait is the culprit.) I'll work around it, if that's a non-issue for you (I'm very new to Rust and don't understand all the trade-offs), but I thought I'd bring it to your attention.

**Comment by avl** [comment]
Had to read up a bit on this error handling thing. 

I agree with you that is it is a bit bad that SavefileError doesn't implement 'std::error::Error'. I would like to fix this. This seems to be an expected side effect of using the 'failure' crate, which is what Savefile itself uses for error handling.

As I understand, the expected way to handle this is to use the .compat() call to convert a failure::Error to a std::error::Error: https://docs.rs/failure/0.1.2/failure/trait.Fail.html#method.compat

I think I would like to migrate away from failure though, a crate like Savefile really should be compatible with the standard way to handle errors in the language.
